Add blog post: Fine tuning CLIP with Remote Sensing (Satellite) images and captions

sujitpal opened the pull request:
## What does this PR do?

This PR proposes to integrate a new blog post fine tuning the OpenAI CLIP transformer model with Remote Sensing images and captions. This model was fine-tuned as part of the Hugging Face Flax/JAX Community week.

## Who can review?

Anyone in the community is free to review the PR.

osanseviero commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,103 @@
+
+Our team, like probably many others, is a distributed one, spanning 12 time zones. Our common thread is that we all belong to the [TWIML Slack Channel](https://twimlai.slack.com/), where we came together based on a shared interest in Artificial Intelligence (AI) and Machine Learning (ML) topics. 
+
+We decided that we would fine tune the [CLIP Network from OpenAI](https://openai.com/blog/clip/) with satellite images and captions from the [RSICD dataset](https://github.com/201528014227051/RSICD_optimal). The CLIP network learns visual concepts by being trained with image and caption pairs in a self-supervised manner, by using text paired with images found across the Internet. During inference ...
```
```suggestion
We fine-tuned the [CLIP Network from OpenAI](https://openai.com/blog/clip/) with satellite images and captions from the [RSICD dataset](https://github.com/201528014227051/RSICD_optimal). The CLIP network learns visual concepts by being trained with image and caption pairs in a self-supervised manner, by using text paired with images found across the Internet. During inference, the model can predict the most relevant image given a text description or the most relevant text description given an image. CLIP is powerful enough to be used in zero-shot manner on everyday images. However, we felt that satellite images were sufficiently different from everyday images that it would be useful to fine-tune CLIP with them. Our intuition turned out to be correct, as the evaluation results (described below) shows.
```

sujitpal replied:
accepted without change

osanseviero commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,103 @@
+
+We decided that we would fine tune the [CLIP Network from OpenAI](https://openai.com/blog/clip/) with satellite images and captions from the [RSICD dataset](https://github.com/201528014227051/RSICD_optimal). The CLIP network learns visual concepts by being trained with image and caption pairs in a self-supervised manner, by using text paired with images found across the Internet. During inference ...
+
+We think of our project more as an "applied" project, in the sense that it is more about providing a useful service than about significantly advancing the technological frontier. Our model can be used in applications to search through large collections of satellite images using text, or automatically find specific features in such images. Both these features can be invaluable for "digital assista ...
```
Couple of comments bellow for this pagraph

```suggestion
The goal of our project was to provide a useful service and demonstrate how to use CLIP for practical use cases. Our model can be used in applications to search through large collections of satellite images using text or automatically find specific features in such images. Image and feature finding can be invaluable for applications where humans have to search through enormous quantities of image data such as helping to deal with natural disasters and climate change.  In a broader sense, it demonstrates that CLIP can be similarly fine-tuned for other domains also, such as Radiography or Pathology. 

You can read about our project on our [project page](https://github.com/arampacha/CLIP-rsicd), download our [trained model](https://huggingface.co/flax-community/clip-rsicd-v2) to use for inference on your own data, or see it in action on our [demo](https://huggingface.co/spaces/sujitpal/clip-rsicd-demo).
```

osanseviero replied:
Maybe it would be good to provide an example or describe a bit more what this means

```
or automatically find specific features in such images
```

osanseviero replied:
`In a broader sense, it demonstrates that CLIP can be similarly fine-tuned for other domains also, such as Radiography or Pathology`

I think this is a great opportunity to mention and link to other projects from the event that used CLIP such as https://huggingface.co/spaces/flax-community/medclip-demo

osanseviero replied:
I wonder if we should have a small section of potential misuses or ethical implications of the work

stevhliu replied:
I also think it'd be good to highlight potential misuse, especially since it has massive implications with regard to military and police surveillance.

sujitpal replied:
incorporated all suggestions

osanseviero commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,103 @@
+
+We think of our project more as an "applied" project, in the sense that it is more about providing a useful service than about significantly advancing the technological frontier. Our model can be used in applications to search through large collections of satellite images using text, or automatically find specific features in such images. Both these features can be invaluable for "digital assista ...
+
+Our project ended up placing third at the event, which was a very pleasant surprise. We are incredibly humbled and gratified that the judges saw the potential of our model. We are very grateful to the organizers for introducing us to Flax/JAX and for providing the resources to create and showcase our model. We are also very thankful to the other participants and organizers for sharing their knowl ...
```
I don't have strong opinion, but I think maybe we can skip this paragraph

sujitpal replied:
removing the paragraph

osanseviero commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,103 @@
+
+Our project ended up placing third at the event, which was a very pleasant surprise. We are incredibly humbled and gratified that the judges saw the potential of our model. We are very grateful to the organizers for introducing us to Flax/JAX and for providing the resources to create and showcase our model. We are also very thankful to the other participants and organizers for sharing their knowl ...
+
+In this post, we describe details of our training and evaluation process, and our plans for future work on this project.
```
At this point you're almost at 25% of the blog post. Maybe this intro could go closer to the beginning.

sujitpal replied:
Moved sentence up to earlier paragraph

osanseviero commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,103 @@
+
+### Dataset
+
+Our model was fine-tuned primarily using the RSICD dataset, which is a set of about 10,000 images collected from Google Earth, Baidu Map, MapABC and Tianditu, and provided freely to the research community for advancement of remote sensing captioning via [Exploring Models and Data for Remote Sensing Image Caption Generation](https://arxiv.org/abs/1712.07835) (Lu et al, 2017). The images are provid ...
```
It would be great if you add a link to the RSICD dataset here

sujitpal replied:
Done, didn't link this previously because it was already linked.

osanseviero commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,103 @@
+### Dataset
+
+Our model was fine-tuned primarily using the RSICD dataset, which is a set of about 10,000 images collected from Google Earth, Baidu Map, MapABC and Tianditu, and provided freely to the research community for advancement of remote sensing captioning via [Exploring Models and Data for Remote Sensing Image Caption Generation](https://arxiv.org/abs/1712.07835) (Lu et al, 2017). The images are provid ...
+
```
I think this is a nice opportunity to put some images of the dataset

sujitpal replied:
done, also added a "header" image for the post from the demo

osanseviero commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,103 @@
+
+Our model was fine-tuned primarily using the RSICD dataset, which is a set of about 10,000 images collected from Google Earth, Baidu Map, MapABC and Tianditu, and provided freely to the research community for advancement of remote sensing captioning via [Exploring Models and Data for Remote Sensing Image Caption Generation](https://arxiv.org/abs/1712.07835) (Lu et al, 2017). The images are provid ...
+
+In addition, we used the [UCM Dataset](https://mega.nz/folder/wCpSzSoS#RXzIlrv--TDt3ENZdKN8JA) and the [Sydney dataset](https://mega.nz/folder/pG4yTYYA#4c4buNFLibryZnlujsrwEQ) for training, The UCM dataset is based on the UC Merced Land Use dataset. It consists of 2100 images belonging to 21 classes (100 images per class). The dataset provides 5 captions for each image. The Sydney dataset contain ...
```
It would be great to understand a bit more why you chose to use those additional datasets

osanseviero commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,103 @@
+
+### Model
+
+Our model is just the fine tuned version of the original CLIP model shown below. Inputs to the model are a batch of captions and a batch of images passed through a text encoder and image encoder respectively. The training process uses contrastive learning to push closer images and their respective captions and to push apart images and captions for other images. In this way, images and captions th ...
```
```suggestion
Our model is just the fine-tuned version of the original CLIP model shown below. Inputs to the model are a batch of captions and a batch of images passed through a text encoder and image encoder respectively. The training process uses contrastive learning to push closer images and their respective captions and to push apart images and captions for other images. In this way, images and captions that belong together get pushed closer together.
```

osanseviero replied:
I would not assume readers know what contrastive learning is. Maybe 1-2 sentences or a link to an external resource would go a long way here.

sujitpal replied:
linked contrastive learning and made explanation more clear

osanseviero commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,103 @@
+
+### Data Augmentation
+
+Because our dataset was not very large, we used both image augmentation and text augmentation in an effort to regularize our dataset and prevent overfitting.
```
```suggestion
In order to regularize our dataset and prevent overfitting due to the size of the dataset, we used image and text augmentation.
```

sujitpal replied:
accepted without change

osanseviero commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,103 @@
+
+Because our dataset was not very large, we used both image augmentation and text augmentation in an effort to regularize our dataset and prevent overfitting.
+
+Image augmentation was done inline using built-in transforms from Pytorch's Torchvision package. The transformations used were Random Cropping, Random Resizing and Cropping, Color Jitter, and Random Horizontal and Vertical flipping.
```
Maybe worth linking to torchvision

sujitpal replied:
linked

osanseviero commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,103 @@
+
+Image augmentation was done inline using built-in transforms from Pytorch's Torchvision package. The transformations used were Random Cropping, Random Resizing and Cropping, Color Jitter, and Random Horizontal and Vertical flipping.
+
+Text augmentation was done, using backtranslation, to generate captions for images that had less than 5 unique captions per image. We used the [Marian MT family of models from Hugging Face](https://huggingface.co/transformers/model_doc/marian.html) to translate the existing captions into French, Spanish, Italian and Portugese and back to English to fill out the captions for these images.
```
```suggestion
Text augmentation was done using backtranslation to generate captions for images that had less than 5 unique captions per image. We used the [Marian MT family of models from Hugging Face](https://huggingface.co/transformers/model_doc/marian.html) to translate the existing captions into French, Spanish, Italian and Portugese and back to English to fill out the captions for these images.
```

sujitpal replied:
accepted without change

osanseviero commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,103 @@
+
+### Metrics
+
+For the evaluation, we used a subset of the RSICD test set where image file names contained the category the image belonged to. We found 30 categories of images in this subset. Evaluation was done by comparing each image with a set of 30 caption sentences of the form `"An aerial photograph of {category}"`. The model produced a ranked list of the 30 captions, from most relevant to least relevant.  ...
```
```suggestion
A subset of the RSICD test set was used for evaluation by . We found 30 categories of images in this subset. The evaluation was done by comparing each image with a set of 30 caption sentences of the form `"An aerial photograph of {category}"`. The model produced a ranked list of the 30 captions, from most relevant to least relevant. Categories corresponding to captions with the top k scores (for k=1, 3, 5, and 10) were compared with the "label" category. The scores are averaged over the entire set of images used for evaluation and reported for various values of k, as shown below.
```

sujitpal replied:
accepted without change (except fix grammatical error)

osanseviero commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,103 @@
+
+For the evaluation, we used a subset of the RSICD test set where image file names contained the category the image belonged to. We found 30 categories of images in this subset. Evaluation was done by comparing each image with a set of 30 caption sentences of the form `"An aerial photograph of {category}"`. The model produced a ranked list of the 30 captions, from most relevant to least relevant.  ...
+
+The `baseline` model represents the pre-trained `openai/clip-vit-base-path32` CLIP model. This model was fine tuned with captions and images from the RSICD dataset, which resulted in a significant performance boost, as shown below.
```
```suggestion
The `baseline` model represents the pre-trained `openai/clip-vit-base-path32` CLIP model. This model was fine-tuned with captions and images from the RSICD dataset, which resulted in a significant performance boost, as shown below.
```

sujitpal replied:
accepted without change

osanseviero commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,103 @@
+We are grateful that we have been given an opportunity to further refine our model. Some ideas we have for future work are as follows:
+
+1. Construct a sequence to sequence model using a CLIP encoder and a GPT-3 decoder and train it to caption images.
+2. Fine tune the model on more image caption pairs from other datasets and investigate if we can improve its performance.
```
```suggestion
2. Fine-tune the model on more image caption pairs from other datasets and investigate if we can improve its performance.
```

sujitpal replied:
accepted without change

osanseviero commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,103 @@
+1. Construct a sequence to sequence model using a CLIP encoder and a GPT-3 decoder and train it to caption images.
+2. Fine tune the model on more image caption pairs from other datasets and investigate if we can improve its performance.
+3. Investigate how fine-tuning affects the performance of model on non-RSICD image caption pairs.
+4. Investigate the capability of the fine-tuned model to classify outside the categories it has been fine tuned on.
```
```suggestion
4. Investigate the capability of the fine-tuned model to classify outside the categories it has been fine-tuned on.
```

sujitpal replied:
accepted without change

osanseviero reviewed (commented):
This looks great!

I left a series of suggestions and small nits but overall things seem fine. One feedback I would have is to make things a bit more visual. Having an image of your demo or an example input/output somewhere in the intro would be nice, that way you excite readers even more!

stevhliu commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,103 @@
+_by Artashes Arutiunian (@arampacha), Dev Vidhani (@devvidhani), Goutham Venkatesh (@goutham794), Mayank Bhaskar (@cataluna84), Ritobrata Ghosh (@ghosh-r), and Sujit Pal (@sujitpal)_
+
+
+In July this year, [Hugging Face](https://huggingface.co/) organized a [Flax/JAX Community Week](https://github.com/huggingface/transformers/blob/master/examples/research_projects/jax-projects/README.md#quickstart-flax-and-jax), during which the community was invited to submit projects to train Hugging Face transformer models in the areas of Natural Language Processing (NLP) and Computer Vision ( ...
```
Small suggestion for using active voice instead of passive voice :)

```suggestion
In July this year, [Hugging Face](https://huggingface.co/) organized a [Flax/JAX Community Week](https://github.com/huggingface/transformers/blob/master/examples/research_projects/jax-projects/README.md#quickstart-flax-and-jax), and invited the community to submit projects to train Hugging Face transformer models in the areas of Natural Language Processing (NLP) and Computer Vision (CV) on Tensor Processing Units (TPU) using [Flax](https://github.com/google/flax), a neural network library and ecosystem for [JAX](https://github.com/google/jax), a linear algebra library (like numpy) which can do automatic differentiation ([Autograd](https://github.com/hips/autograd)) and can compile down to [XLA](https://www.tensorflow.org/xla). [Google Cloud](https://cloud.google.com/) were co-sponsors of this event.
```

sujitpal replied:
done

stevhliu commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,103 @@
+
+We decided that we would fine tune the [CLIP Network from OpenAI](https://openai.com/blog/clip/) with satellite images and captions from the [RSICD dataset](https://github.com/201528014227051/RSICD_optimal). The CLIP network learns visual concepts by being trained with image and caption pairs in a self-supervised manner, by using text paired with images found across the Internet. During inference ...
+
+We think of our project more as an "applied" project, in the sense that it is more about providing a useful service than about significantly advancing the technological frontier. Our model can be used in applications to search through large collections of satellite images using text, or automatically find specific features in such images. Both these features can be invaluable for "digital assista ...
```
This paragraph was a little hard to follow, so I just made some minor grammatical changes. Feel free to use this or @osanseviero's suggestion to break up the paragraph into two shorter ones :)

```suggestion
We think of our project more like an "applied" project because it is more about providing a useful service than significantly advancing the technological frontier. Our model can be used in applications to search through large collections of satellite images using text, or automatically find specific features in such images. These features can be invaluable for "digital assistant" type applications ranging from military/police surveillance to helping deal with natural disasters and climate change -- any application where humans have to search through enormous quantities of image data. In a broader sense, our project demonstrates that CLIP can also be similarly fine-tuned for other domains, such as Radiography or Pathology. You can read about our project on our [project page](https://github.com/arampacha/CLIP-rsicd), download our [trained model](https://huggingface.co/flax-community/clip-rsicd-v2) for inference on your data, or see it in action on our [demo](https://huggingface.co/spaces/sujitpal/clip-rsicd-demo).
```

sujitpal replied:
Following @osanseviero 's feedback, we removed many unnecessary details from this paragraph and so resulting shortened version is no longer large and hard to read.

stevhliu commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,103 @@
+
+### Dataset
+
+Our model was fine-tuned primarily using the RSICD dataset, which is a set of about 10,000 images collected from Google Earth, Baidu Map, MapABC and Tianditu, and provided freely to the research community for advancement of remote sensing captioning via [Exploring Models and Data for Remote Sensing Image Caption Generation](https://arxiv.org/abs/1712.07835) (Lu et al, 2017). The images are provid ...
```
I think this would be easier to follow if it was broken up into several shorter sentences.

```suggestion
We fine-tuned our model primarily with the RSICD dataset. This dataset consists of about 10,000 images collected from Google Earth, Baidu Map, MapABC, and Tianditu. It is provided freely to the research community to advance remote sensing captioning via Exploring Models and Data for Remote Sensing Image Caption Generation (Lu et al, 2017). The images are (224, 224) RGB images at various resolutions, and each image has up to 5 captions associated with it.
```

sujitpal replied:
accepted without change, put back links from previous version suggested by @osanseviero

stevhliu commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,103 @@
+
+Our model was fine-tuned primarily using the RSICD dataset, which is a set of about 10,000 images collected from Google Earth, Baidu Map, MapABC and Tianditu, and provided freely to the research community for advancement of remote sensing captioning via [Exploring Models and Data for Remote Sensing Image Caption Generation](https://arxiv.org/abs/1712.07835) (Lu et al, 2017). The images are provid ...
+
+In addition, we used the [UCM Dataset](https://mega.nz/folder/wCpSzSoS#RXzIlrv--TDt3ENZdKN8JA) and the [Sydney dataset](https://mega.nz/folder/pG4yTYYA#4c4buNFLibryZnlujsrwEQ) for training, The UCM dataset is based on the UC Merced Land Use dataset. It consists of 2100 images belonging to 21 classes (100 images per class). The dataset provides 5 captions for each image. The Sydney dataset contain ...
```
```suggestion
In addition, we used the [UCM Dataset]((https://mega.nz/folder/wCpSzSoS#RXzIlrv--TDt3ENZdKN8JA)) and the [Sydney dataset](https://mega.nz/folder/pG4yTYYA#4c4buNFLibryZnlujsrwEQ) for training. The UCM dataset is based on the UC Merced Land Use dataset. It consists of 2100 images belonging to 21 classes (100 images per class), and each image has 5 captions. The Sydney dataset contains images of Sydney, Australia, from Google Earth. It contains 613 images belonging to 7 classes. Images are (500, 500) RGB and provide 5 captions for each image.
```

sujitpal replied:
accepted without change

stevhliu commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,103 @@
+
+### Model
+
+Our model is just the fine tuned version of the original CLIP model shown below. Inputs to the model are a batch of captions and a batch of images passed through a text encoder and image encoder respectively. The training process uses contrastive learning to push closer images and their respective captions and to push apart images and captions for other images. In this way, images and captions th ...
```
This paragraph was also a little hard to follow. When you say **closer images**, it's not clear if you are pushing images that are closer together, or if you are just pushing the images closer.

```suggestion
Our model is just the fine-tuned version of the original CLIP model shown below. Inputs to the model are a batch of captions and images passed through a text encoder and image encoder respectively. The training process uses contrastive learning to push similar images and their respective captions together. Dissimilar images and captions get pushed apart.
```

sujitpal replied:
I have reworded it to link contrastive learning and to provide little more background. Is this understandable or should I reword further?

stevhliu commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,103 @@
+
+Image augmentation was done inline using built-in transforms from Pytorch's Torchvision package. The transformations used were Random Cropping, Random Resizing and Cropping, Color Jitter, and Random Horizontal and Vertical flipping.
+
+Text augmentation was done, using backtranslation, to generate captions for images that had less than 5 unique captions per image. We used the [Marian MT family of models from Hugging Face](https://huggingface.co/transformers/model_doc/marian.html) to translate the existing captions into French, Spanish, Italian and Portugese and back to English to fill out the captions for these images.
```
```suggestion
We augmented the text with backtranslation to generate captions for images with less than 5 unique captions per image. The [Marian MT]((https://huggingface.co/transformers/model_doc/marian.html)) family of models from Hugging Face was used to translate the existing captions into French, Spanish, Italian, and Portuguese and back to English to fill out the captions for these images.
```

sujitpal replied:
accepted without change

stevhliu commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,103 @@
+
+The `baseline` model represents the pre-trained `openai/clip-vit-base-path32` CLIP model. This model was fine tuned with captions and images from the RSICD dataset, which resulted in a significant performance boost, as shown below.
+
+Our best model was trained with image and text augmentation, with batch size 1024 (128 spread across 8 TPU devices), and the Adam optimizer with learning rate 5e-6. Our second base model was trained with the same hyperparameters, except that we used the Adafactor optimizer with learning rate 1e-4. You can download either model from their model cards linked to in the table below.
```
```suggestion
Our best model was trained with image and text augmentation, with batch size 1024 (128 spread across 8 TPU devices), and the Adam optimizer with learning rate 5e-6. We trained our second base model with the same hyperparameters, except that we used the Adafactor optimizer with learning rate 1e-4. You can download either model from their model cards linked to in the table below.
```

sujitpal replied:
accepted without change

stevhliu commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,103 @@
+
+### Demo
+
+You can access our [CLIP-RSICD Demo](https://huggingface.co/spaces/sujitpal/clip-rsicd-demo) here. It uses our fine-tuned CLIP model to provide the following functionality. 
```
```suggestion
You can access our [CLIP-RSICD Demo](https://huggingface.co/spaces/sujitpal/clip-rsicd-demo) here. It uses our fine-tuned CLIP model to provide the following functionality:
```

sujitpal replied:
accepted without change

stevhliu commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,103 @@
+* Image to Image search
+* Find text feature in image
+
+For the first two features, the "image corpus" is composed of images from the RSICD test set. They are encoded using our best fine-tuned CLIP model and stored in a NMSLib index which allows Approximate Nearest Neighbor based retrieval. For text-to-image and image-to-image search respectively, the query text or image are encoded with our model and matched against the image vectors in the corpus. F ...
```
```suggestion
The first two functionalities uses the "image corpus" from the RSICD test set. They are encoded using our best fine-tuned CLIP model and stored in a NMSLib index which allows Approximate Nearest Neighbor based retrieval. For text-to-image and image-to-image search respectively, the query text or image are encoded with our model and matched against the image vectors in the corpus. For the third functionality, we partition the incoming image into patches and encode them, encode the queried text feature, match the text vector with each image patch vector, and return the probability of finding the feature in each patch.
```

sujitpal replied:
accepted without change except NMSLib linked

sujitpal commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,103 @@
+
+# Fine tuning CLIP with Remote Sensing (Satellite) images and captions
+
+_by Artashes Arutiunian (@arampacha), Dev Vidhani (@devvidhani), Goutham Venkatesh (@goutham794), Mayank Bhaskar (@cataluna84), Ritobrata Ghosh (@ghosh-r), and Sujit Pal (@sujitpal)_
```
Done, awaiting gravatar thumbnails if applicable

sujitpal commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,103 @@
+_by Artashes Arutiunian (@arampacha), Dev Vidhani (@devvidhani), Goutham Venkatesh (@goutham794), Mayank Bhaskar (@cataluna84), Ritobrata Ghosh (@ghosh-r), and Sujit Pal (@sujitpal)_
+
+
+In July this year, [Hugging Face](https://huggingface.co/) organized a [Flax/JAX Community Week](https://github.com/huggingface/transformers/blob/master/examples/research_projects/jax-projects/README.md#quickstart-flax-and-jax), during which the community was invited to submit projects to train Hugging Face transformer models in the areas of Natural Language Processing (NLP) and Computer Vision ( ...
```
Accepted with changes to ensure continuity

sujitpal commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,103 @@
+
+In July this year, [Hugging Face](https://huggingface.co/) organized a [Flax/JAX Community Week](https://github.com/huggingface/transformers/blob/master/examples/research_projects/jax-projects/README.md#quickstart-flax-and-jax), during which the community was invited to submit projects to train Hugging Face transformer models in the areas of Natural Language Processing (NLP) and Computer Vision ( ...
+
+They organized 3 days of lectures during which speakers from Hugging Face and Google Cloud talked about TPUs, JAX, Flax, and how to use them to train Hugging Face transformer models (recordings for [day #1](https://www.youtube.com/watch?v=fuAyUQcVzTY), [day #2](https://www.youtube.com/watch?v=__eG63ZP_5g), and [day #3](https://www.youtube.com/watch?v=ZCMOPkcTu3s) available). This was followed by  ...
```
Accepted with some minor changes.

sujitpal commented:
Thank you for your feedback @osanseviero and @stevhliu . For those suggestions that I accepted verbatim I have annotated with "accepted without change" and others I have taken the idea and expanded on it slightly differently. I have committed changes based on your feedback. Would appreciate another round of review to see if this is good.

Also I have asked my co-authors for their image profile image links, if I get back everyone's I will put them in to the header, otherwise we will stick to the current text only format.

osanseviero reviewed (commented):
This looks great. Thanks for doing all the changes. I'll let @stevhliu and @patil-suraj take an additional look, but imo this is pretty much ready.

For the intro gravatars, we usually use the hf.co profile gravatar (and username). You can right-click and get the url.

stevhliu commented:
Looks good to me. Thanks again for this interesting project and blog post! :)

sujitpal commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,163 @@
+
+The goal of our project was to provide a useful service and demonstrate how to use CLIP for practical use cases. Our model can be used by applications to search through large collections of satellite images using textual queries. Such queries could describe the image in totality (for example, beach, mountain, airport, baseball field, etc) or search or mention specific geographic or man-made featu ...
+
+The ability to search through large collections of images using text queries is an immensely powerful feature, and can be used as much for social good as for malign purposes. Possible applications include national defense and anti-terrorism activities, and the ability to spot and address effects of climate change before they become unmanageable, etc. Unfortunately, this power can also be misused, ...
```
👍

sujitpal commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,163 @@
+
+Participants used Tensor Processing Units (TPUs) with [Flax](https://github.com/google/flax) and [JAX](https://github.com/google/jax). JAX is a linear algebra library (like `numpy`) that can do automatic differentiation ([Autograd](https://github.com/hips/autograd)) and compile down to [XLA](https://www.tensorflow.org/xla), and Flax is a neural network library and ecosystem for JAX. TPU compute t ...
+
+Over the next two weeks, teams participated in lectures from Hugging Face and Google, trained one or more models using JAX/Flax, shared them with the community, and provided a a demo with [Hugging Face spaces](https://huggingface.co/spaces) showcasing the capabilities of their model.  Approximately 100 teams participated in the event, and it resulted in 170 models and 36 demos.
```
accepted without change

sujitpal commented:
I have updated with latest feedback, also thanks for the pointer to HF profiles, I have updated and asked 3 of my teammates to update their photographs, will update the links once they are done, and address any further feedback from @patil-suraj

osanseviero commented:
Hi @sujitpal! I think we can make this PR ready and merge it. 


* We need to change the directory name of the assets since there are a couple of new blog posts.
* We'll also need to add an entry to https://github.com/huggingface/blog/blob/master/_blog.yml
* Finally we need to change the date in the blog post

Once those 3 things are done I can merge and publish the blog post!

patil-suraj commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,169 @@
+
+The `baseline` model represents the pre-trained `openai/clip-vit-base-path32` CLIP model. This model was fine-tuned with captions and images from the RSICD dataset, which resulted in a significant performance boost, as shown below.
+
+Our best model was trained with image and text augmentation, with batch size 1024 (128 spread across 8 TPU devices), and the Adam optimizer with learning rate 5e-6. We trained our second base model with the same hyperparameters, except that we used the Adafactor optimizer with learning rate 1e-4. You can download either model from their model cards linked to in the table below.
```
```suggestion
Our best model was trained with image and text augmentation, with batch size 1024 (128 on each of the 8 TPU cores), and the Adam optimizer with learning rate 5e-6. We trained our second base model with the same hyperparameters, except that we used the Adafactor optimizer with learning rate 1e-4. You can download either model from their model repos linked to in the table below.
```

sujitpal replied:
applied w/o modifications

patil-suraj commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,169 @@
+
+#### Demo
+
+You can access our [CLIP-RSICD Demo](https://huggingface.co/spaces/sujitpal/clip-rsicd-demo) here. It uses our fine-tuned CLIP model to provide the following functionality:
```
(nit)
```suggestion
You can access the [CLIP-RSICD Demo](https://huggingface.co/spaces/sujitpal/clip-rsicd-demo) here. It uses our fine-tuned CLIP model to provide the following functionality:
```

patil-suraj commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,169 @@
+* Image to Image search
+* Find text feature in image
+
+The first two functionalities uses the RSICD test set as it's image corpus. They are encoded using our best fine-tuned CLIP model and stored in a [NMSLib](https://github.com/nmslib/nmslib) index which allows Approximate Nearest Neighbor based retrieval. For text-to-image and image-to-image search respectively, the query text or image are encoded with our model and matched against the image vector ...
```
(nit)
```suggestion
The first two functionalities use the RSICD test set as its image corpus. They are encoded using our best fine-tuned CLIP model and stored in a [NMSLib](https://github.com/nmslib/nmslib) index which allows Approximate Nearest Neighbor based retrieval. For text-to-image and image-to-image search respectively, the query text or image are encoded with our model and matched against the image vectors in the corpus. For the third functionality, we divide the incoming image into patches and encode them, encode the queried text feature, match the text vector with each image patch vector, and return the probability of finding the feature in each patch.
```

patil-suraj commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,169 @@
+
+#### Dataset
+
+We fine-tuned our model primarily with the [RSICD dataset](https://github.com/201528014227051/RSICD_optimal). This dataset consists of about 10,000 images collected from Google Earth, Baidu Map, MapABC, and Tianditu. It is provided freely to the research community to advance remote sensing captioning via [Exploring Models and Data for Remote Sensing Image Caption Generation](https://arxiv.org/abs ...
```
> The images are (224, 224) RGB images at various resolutions

This sounds a bit confusing, the resolution is specified here (224, 224) but then it also says  various resolutions

patil-suraj replied:
```suggestion
We fine-tuned the CLIP model primarily with the [RSICD dataset](https://github.com/201528014227051/RSICD_optimal). This dataset consists of about 10,000 images collected from Google Earth, Baidu Map, MapABC, and Tianditu. It is provided freely to the research community to advance remote sensing captioning via [Exploring Models and Data for Remote Sensing Image Caption Generation](https://arxiv.org/abs/1712.0783) (Lu et al, 2017). The images are (224, 224) RGB images at various resolutions, and each image has up to 5 captions associated with it.
```

patil-suraj commented on fine-tune-clip-rsicd.md:
```diff
@@ -0,0 +1,169 @@
+
+The ability to search through large collections of images using text queries is an immensely powerful feature, and can be used as much for social good as for malign purposes. Possible applications include national defense and anti-terrorism activities, the ability to spot and address effects of climate change before they become unmanageable, etc. Unfortunately, this power can also be misused, suc ...
+
+You can read about our project on our [project page](https://github.com/arampacha/CLIP-rsicd), download our [trained model](https://huggingface.co/flax-community/clip-rsicd-v2) to use for inference on your own data, or see it in action on our [demo](https://huggingface.co/spaces/sujitpal/clip-rsicd-demo).
```
(nit)
```suggestion
You can read about the project on our [project page](https://github.com/arampacha/CLIP-rsicd), download the [trained model](https://huggingface.co/flax-community/clip-rsicd-v2) to use for inference on your own data, or see it in action on our [demo](https://huggingface.co/spaces/sujitpal/clip-rsicd-demo).
```

patil-suraj reviewed (approved):
Sorry about the super late review.

This looks good, great job! I just left few tiny nits. :)
